### PR TITLE
Add SCAI reusable actions (KubeCon NA '23 Demo)

### PR DESCRIPTION
**Pull request by marcelamelara**
This PR adds GitHub Actions for running the Go `scai-gen` tools through GHA workflows. This PR also adds a new (experimental) command `scai-gen sigstore` for Sigstore integration (cosign signing and Rekor upload) via GHA. The primary use of these Actions is as part of the in-toto KubeCon NA '23 demo. The demo implementation lives [here](https://github.com/marcelamelara/private-data-objects/blob/kubeconNA23-intoto-demo/.github/workflows/kubeconNA23-intoto-demo.yml).

### CODE REVIEW

File: .github/workflows/test-e2e-flow.yml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+      - name: Install scai-gen CLI tools
+        shell: bash
+        run: |
+          go install ./scai-gen
+          mkdir -p temp
+
+      - name: Generate SBOM SCAI AttributeAssertion
+        id: gen-sbom-assert
+        uses: marcelamelara/scai-demos/.github/actions/scai-gen-assert@add-scai-reusable-workflows
```
Comment (adityasaky) [inline .github/workflows/test-e2e-flow.yml:38]: hm, do these get bumped up in a separate PR?

Comment (adityasaky) [reply .github/workflows/test-e2e-flow.yml:38]: I wonder if you could `uses: ./.github/actions` etc. https://github.com/gittuf/gittuf-installer/blob/main/.github/workflows/test.yml#L29

I'm not sure though.

File: scai-gen/cmd/sigstore.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	// will only sign valid Statements
+	err = statement.Validate()
+	if err != nil {
+		return fmt.Errorf("invalid in-toto Statement: %w", err)
+	}
+
+	ctx := context.Background()
+
+	// Get Fulcio signer
+	if !providers.Enabled(ctx) {
```
Comment (adityasaky) [inline scai-gen/cmd/sigstore.go:92]: I'd re-enable cmd in the args and use `cmd.Context()`. If one isn't set, it eventually hits `Background` too.

**Review by adityasaky — COMMENTED** [review]
I'm not an expert in actions and parts of sigstore but some quick thoughts.

File: .github/workflows/test-e2e-flow.yml (hunk 1 above)
Comment (marcelamelara) [reply .github/workflows/test-e2e-flow.yml:38]: I've been trying to figure this out, too. I'll try your suggestion with relative paths, thanks!

File: scai-gen/cmd/sigstore.go (hunk 2 above)
Comment (marcelamelara) [reply scai-gen/cmd/sigstore.go:92]: Just to be 100% sure I understood the change you're suggesting. I should replace the `providers.Enabled(ctx)` check with `providers.Enabled(cmd.Context())?

Comment (adityasaky) [reply scai-gen/cmd/sigstore.go:92]: Yes: https://pkg.go.dev/github.com/spf13/cobra#Command.Context

File: scai-gen/cmd/sigstore.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+	dsseSigner := dsse.WrapSigner(k, "application/vnd.in-toto")
+	signedAtt, err := dsseSigner.SignMessage(bytes.NewReader(attBytes))
+	if err != nil {
+		return fmt.Errorf("error signing DSSE: %w", err)
+	}
+
+	// Add certificate to envelope. This is needed for
+	// Rekor compatibility.
+	signedAttWithCert, err := envelope.AddCertToEnvelope(signedAtt, k.Cert)
```
Comment (adityasaky) [inline scai-gen/cmd/sigstore.go:114]: This is technically not part of the DSSE spec. I don't want to block this PR on this but this is another data point for us to get https://github.com/secure-systems-lab/dsse/pull/61 merged and implemented.

**Review by adityasaky — APPROVED** [review]
Mostly nits but I think at this point this is fine.

File: .github/actions/scai-gen-assert/action.yml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        name: "${{ inputs.evidence-file }}"
+        path: "${{ inputs.evidence-path }}"
+        media-type: "${{ inputs.evidence-type }}"
+        rd-name: "${{ inputs.evidence-file }}-desc.json"
+
+    - name: Run scai-gen assert
+      id: scai-gen-assert
+      shell: bash
+      run: |
+        mkdir -p ${{ inputs.assertion-path }}
```
[the remark this answers is not in the document]
Comment (marcelamelara) [reply .github/actions/scai-gen-assert/action.yml:57]: This is a good point, and shouldn't be too hard to implement, I hope. What I'll do is merge this PR so I can start testing this with my test build. I have some documentation for the demo in-flight (nothing major) that I think would fit well into a "make the demo more legible/usable" PR.

File: scai-gen/cmd/sigstore.go (hunk 3 above)
Comment (marcelamelara) [reply scai-gen/cmd/sigstore.go:114]: 100%

File: .github/actions/scai-gen-assert/action.yml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        scai-gen assert -e ${{ steps.gen-rd.outputs.file-rd-name }} -o ${{ inputs.assertion-path }}/${{ inputs.assertion-name }} ${{ inputs.attribute}}
```
[the remark this answers is not in the document]
Comment (marcelamelara) [reply .github/actions/scai-gen-assert/action.yml:58]: Thanks! I'll make these changes as part of a follow-up PR that focuses on readability and usability of the demo, so I can unblock testing on the implementer end.

**Comment by adityasaky** [comment]
Sounds good, I'm fine with this being merged now.

**Pull request merged**